Query level nullability review and drafting thread (#1)
twof/graphql-spec: rfcs/QueryLevelNullability.md

# RFC: Query Level Nullability 

**Proposed by:** 
- [Liz Jakubowski](<social or github link here>) - Yelp iOS

> - [Alex Reilly](<social or github link here>) - Yelp iOS

Author: Example of a suggested change

```suggestion
- [Alex Reilly](https://github.com/twof/) - Yelp iOS
```

a hand-written business object with a non-nullable `name` property. From that point on, all feature code
can happily assume that it has a non-null business name to work with.

## 🧑‍💻 Proposed syntax

It would make more sense if the client could express that `name` must be non-null _in the query itself_:
```graphql
query GetBusinessName($encid: String!) {
  business(encid: $encid) {
    name!  <-- this!
  }
}
```
Semantically the GraphQL `!` operator is nearly identical to its counterpart in Swift (also represented by `!`) which is
referred to as the "force unwrap operator". In Swift, for example, you can cast a string to an integer with `Int("5")` 
but the string being cast may not be a valid number, so that statement will return `null` rather than an integer if the
string cannot be turned into an integer. If you want to ensure that the statement does not return `null` you can instead 
write `Int("5")!`. If you do that, an exception will be thrown if the statement would return `null`.

In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an

> error to the client rather than any data.

Author: What kind of error is best in this situation? Should that be at the discretion of the user rather than being baked into the spec?

Reviewer: if it were at the discretion of the user, I think it would need to be specified somehow in the syntax. is there another way the user could specify this?

I don't mind a generic `missingRequiredField` error or something along those lines -- what error is returned for the `@nonNull` directive?

# RFC: Query Level Nullability 

**Proposed by:** 

> - [Liz Jakubowski](<social or github link here>) - Yelp iOS

Reviewer: ```suggestion
- [Liz Jakubowski](https://github.com/lizjakubowski) - Yelp iOS
```

- [Sanae Rosen](<social or github link here>) - Yelp Android
- [Mark Larah](https://github.com/magicmark) - Yelp Web

> - [Alex Reilly](<social or github link here>) - Yelp iOS

Reviewer: ```suggestion
- [Alex Reilly](<social or github link here>) - Yelp iOS
- [Wei Xue](<social or github link here>) - Yelp iOS
```

This RFC proposes creating a syntactical construct for client developers to 

> express "nullability" in their queries.

Reviewer: ```suggestion
This RFC proposes a syntactical construct for GraphQL clients to express the **nullability** of schema fields requested in a query.
```

## Definitions

Nullability: A concept that exists across many programming languages (eg [Swift](https://developer.apple.com/documentation/swift/optional), [Kotlin](https://kotlinlang.org/docs/null-safety.html#nullable-types-and-non-null-types), [SQL](https://www.w3schools.com/sql/sql_notnull.asp))

> that is used to express when users can be certain that a value can or can never be `null` 

Reviewer: nit: we should use one term to refer to `(users | developers | clients)` throughout the RFC

Author: Decided to use `developers` throughout

(or the language equivalent). Nullability language constructs (eg `?` in Swift/Kotlin)
have become popular due to their ability to solve ergonomic problems in languages
such as those surrounding `NullPointerException` in Java.

Codegen: Short for Code Generation, tools that exist for working with GraphQL which take queries and schemas as
input and output code in a language of choice that represents data that will result from those 

> queries. Codegen tools exist for many platforms. As an example, [here's some info about Apollo's offerings.](https://github.com/apollographql/apollo-tooling#code-generation)

Reviewer: I replaced "queries" with "documents" ([spec](https://spec.graphql.org/June2018/#sec-Language.Document)), and rephased the second half a bit -- lmk whatcha think!
```suggestion
Codegen: Short for Code Generation, tools that exist for working with GraphQL which accept a schema and a set of documents as
input, and output code in a language of choice that represents the data returned by those 
operations. GraphQL codegen tools exist for many platforms. As an example, [here's some info about Apollo's offerings](https://github.com/apollographql/apollo-tooling#code-generation).
```

## 📜 Problem Statement

GraphQL is a "nullable by default" language meaning that all properties are allowed to be `null`.
This is in contrast to the two modern languages used on mobile clients, Swift and Kotlin,
which are both non-null by default. In Swift and Kotlin, unless developers otherwise
specify it, properties cannot be `null`.

This mismatch creates some dissonance for developers who are currently forced into dealing the
problems that commonly surround nullablility in codebases that otherwise do not need to deal with

> those problems. This makes large numbers of nulls tedious and time-consuming to handle.

Reviewer: ```suggestion
This mismatch creates some dissonance for developers who are currently forced to deal with
problems that commonly surround nullablility. This makes large numbers of null fields tedious and
time-consuming to work with.
```

It is often unclear how to handle partial results.
- What elements are essential to having an adequate user experience? How many elements can fail before you 
  should just show an error message instead?
- When any combination of elements can fail, it is very hard to anticipate all edge cases and how the app 

>   might look and work, including transitions to other screens.

Reviewer: let's say a little more about how this ties in to the proposal! partial results are definitely related to this problem, but they don't go away if the proposal is accepted. maybe what we want to get across is that the proposed syntax changes where these decisions are made, and removes boilerplate from the client?

nitnit: we should use a word other than "app" to keep this as platform-agnostic as possible

In Yelp's GraphQL schema, almost all object fields are nullable except for those with ID type. 

> This adheres to what seems to be the [official best practice](https://graphql.org/learn/best-practices/#nullability).

Reviewer: this is a lil redundant with the first part of the problem statement, maybe we can combine them somehow and remove the Yelp-specific aspect of it. e.g. could just add `According to the [official best practice]() ...` to the original problem statement or something like that :)

> This poses a problem for the mobile clients that use Apollo's codegen feature. The codegen provides 

Reviewer: I think I wrote this originally in the JIRA ticket, where it made more sense to name-drop Apollo in the problem statement. for the RFC, I think we should use Apollo only to illustrate concepts -- the problem applies to any GraphQL code generator that outputs result/model types

we could emphasize that any code generator would use the schema to determine whether the properties on the generated models are optional. if the best practice is to make all schema fields nullable, it follows that the properties on the generated models will be optional
```suggestion
This poses a problem for GraphQL clients that use codegen. The codegen provides 
```
similarly, for the RFC we should make sure that the problem is posed as a cross-platform issue and audit any references I made to mobile-specific stuff in the original JIRA ticket

Swift/Kotlin types to represent queries used in the app. Nullable fields in the schema are represented
as nullable properties on the resulting type, represented by `?` following the type name:
```graphql
query GetBusinessName($encid: String!) {
  business(encid: $encid) {
    name
  }
}
```
```swift
struct GetBusinessNameQuery.Data.Business {
  let name: String?
}
```

> In many cases, the client should error if the business `name` is null. If codegen were out of the picture,

Reviewer 2: It took me several rereads to understand the described use case. Is this a correct understanding? 

While the schema can have the nullable fields for valid reasons (such as federation), the client wants to decide if it accepts a null value for the result to simplify the client-side logic.

Author: Yeah that's right! I took your wording since it's more generally applicable than what we had.

a hand-written business object with a non-nullable `name` property. From that point on, all feature code
can happily assume that it has a non-null business name to work with.

## 🧑‍💻 Proposed syntax

It would make more sense if the client could express that `name` must be non-null _in the query itself_:
```graphql
query GetBusinessName($encid: String!) {
  business(encid: $encid) {
    name!  <-- this!
  }
}
```
Semantically the GraphQL `!` operator is nearly identical to its counterpart in Swift (also represented by `!`) which is
referred to as the "force unwrap operator". In Swift, for example, you can cast a string to an integer with `Int("5")` 
but the string being cast may not be a valid number, so that statement will return `null` rather than an integer if the
string cannot be turned into an integer. If you want to ensure that the statement does not return `null` you can instead 
write `Int("5")!`. If you do that, an exception will be thrown if the statement would return `null`.

In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an
error to the client rather than any data.

> On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.

Reviewer 2: Typescript has a number of codegen tools such as Apollo Codegen and [GraphQL Codegen by The Guild](https://www.graphql-code-generator.com/).

Author: Thanks! This was Yelp specific. We removed it.

On mobile platforms where codegen is used, clients have full control over the nullability of the properties on the
generated types. Since nullability is expressed in the query rather than the schema, it's flexible enough to accommodate

> various use-cases (e.g., where the business `name` _is_ allowed to be nullable).

Reviewer: we should nix the web / mobile distinction for the real RFC, and instead phrase it like "without codegen" and "with codegen"

I don't remember the point I was trying to make with this sentence, it probably makes sense to rework this section a bit
> On web where codegen is not used, the client no longer needs to handle the case where expected fields are missing.

### `!`

We have chosen `!` because `!` is already being used in the GraphQL spec to indicate that a field is non-nullable.
Incidentally the same precedent exists in Swift (`!`) and Kotlin (`!!`) which both use similar syntax to indicate
"throw an exception if this value is `null`". 

### Use cases

#### When a field is necessary to the function of the client
There are times where a field can be `null`, but a feature that queries for the field will not function
if it is `null`. For example if you are trying to render an information page for a movie, you won't
be able to do that if the name field of the movie is missing. In that case it would be preferable

> to fail as early as possible.

Reviewer: is this a Netflix example?

### ✨ Examples

#### Non-nullable field
```graphql
query GetBusinessName($encid: String!) {
  business(encid: $encid) {

>     name!  <-- this!

Reviewer: ```suggestion
    name!
```

  }
}
```
would codegen to the following type on iOS.

```swift
struct GetBusinessNameQuery.Data.Business {
  let name: String // lack of `?` indicates that `name` will never be `null`
}
```

#### Non-nullable object with nullable fields
```graphql
query {
  business(id: 4) {
    reviews! {
      rating
      text
    }
  }
}
```

## ✅ RFC Goals
- Non-nullable syntax that is based off of syntax that developers will already be familiar with

## 🚫 RFC Non-goals

## 🗳️ Alternatives considered

### Make Schema Fields Non-Nullable Instead

> Discussion on [this topic can be found here](https://medium.com/@calebmer/when-to-use-graphql-non-null-fields-4059337f6fc8)

Author: TODO: Flesh this out with things about versioning, breaking changes, and the impact it has on partial results

// We definitely want to flesh this out

### Use a directive such as `@non-null` instead
This is the alternative being used at some of the companies represented in this proposal for the time being.
However this feels like a common enough need to call for a language feature. A single language feature

> rather than using directives as a workaround would enable more unified public tooling around GraphQL.

Reviewer: I think this should be the #1 alternative, since it's probably what folks reviewing will be quick to suggest, given that making schema fields non-nullable (the current #1) is not best practice. whatcha think?

### Write wrapper types that null-check fields
This is the alternative being used at some of the companies represented in this proposal for the time being.
It's quite labor intensive and the work is quite rote. It more or less undermines the purpose of
having code generation.

### Alternatives to `!`
#### `!!`
This would follow the precedent set by Kotlin.

### Make non-nullability apply recursively
For example, everything in this tree would be non-nullable
```graphql
query! {
  business(id: 4) {
    name
  }
}

> ```

Author: @wxue Do you remember why we didn't want this?

Reviewer: I think we should remove this as an alternative, I don't think we understand it well enough to vouch for it

> // TODO: This was rejected when it was proposed. Looking for more info as to why

Reviewer: are you ready for some super meta feedback? 😄 let's remove this code comment and leave a GitHub comment with the TODO instead (so that it's more prominent & promotes discussion)

  ```graphql
  # Schema
  type Business {
    # The unique identifier for the business (non-nullable)
    id: String!
  
    # The name of the business (nullable)
    name: String
  }

  # Document
  query GetBusinessName($id: String!) {
    business(id: $id) {
      name
    }
  }
  ```
  At build time, Apollo generates the following Swift code (note: the code has been shortened for clarity).
  ```swift
  struct GetBusinessNameQuery {
    let id: String

    struct Data {
      let business: Business?

      struct Business {
        /// Since the `Business.name` schema field is nullable, the corresponding codegen Swift property is `Optional`
        let name: String?
      }
    }
  }
  ```
  The query can then be fetched, and the resulting data handled, as follows:
  ```swift
  GraphQL.fetch(query: GetBusinessNameQuery(id: "foo"), completion: { result in
    guard case let .success(gqlResult) = result, let business = gqlResult.data?.business else { return }

    // Often, the client needs to provide a default value in case `name` is `null`.
    print(business?.name ?? "null")
  }
  ```

## 📜 Problem Statement

> The two modern languages used on mobile clients, Swift and Kotlin, are both non-null by default. By contrast, in a

Reviewer 3: I would consider reducing the focus on specific client languages.

The crux of the problem is that the SDL nonNull (!) eliminates the possibility of partial failure on a given type. This forces the SDL author to decide for which fields partial failure is acceptable. A GraphQL schema author may not be in the best position to decide whether partial failure is acceptable for a given canvas. Additionally, the answer to whether a missing value is acceptable may vary between UI canvases. Specifying nonNull (aka. required) in the _query_ gives the UI developer the power to decide where partial failure is acceptable, and the flexibility for different canvases to have different requirements in this regard.

The pains are more noticeable in languages with first-class null safety, but are relevant to any client.

GraphQL type system, every field is nullable by default. From the [official GraphQL best practice](https://graphql.org/learn/best-practices/#nullability):

> This is because there are many things that can go awry in a networked service backed by databases and other
> services. A database could go down, an asynchronous action could fail, an exception could be thrown. Beyond
> simply system failures, authorization can often be granular, where individual fields within a request can
> have different authorization rules.

This discrepancy means that developers using strongly-typed languages must perform `null` checks anywhere that GraphQL
codegen types are used, significantly decreasing the benefits of codegen. In some cases, the codegen types may be so
complex that a new model type which encapsulates the `null` handling is written manually.

While the schema can have nullable fields for valid reasons (such as federation), in some cases the client wants to decide
if it accepts a `null` value for the result to simplify the client-side logic. In addition, syntax for this concept

> would allow codegen tooling to generate model types that are more ergonomic to work with, since the since the model

Reviewer 3: ```suggestion
would allow codegen tooling to generate model types that are more ergonomic to work with, since the model
```

type's properties would have the desired nullability.

## 🧑‍💻 Proposed syntax

The client can express that a schema field is required by using the `!` syntax in the query definition:
```graphql
query GetBusinessName($id: String!) {
  business(id: $id) {
    name!
  }
}
```
Semantically the GraphQL `!` operator is nearly identical to its counterpart in Swift (also represented by `!`) which is
referred to as the "force unwrap operator".

In Swift, for example, you can cast a string to an integer with `Int("5")` but the string being cast may not be a valid
number, so that statement will evaluate to `null` rather than an integer if the string cannot be cast to an integer. If
you want to ensure that the statement does not return `null` you can instead write `Int("5")!`. If you do that, an
exception will be thrown if the statement would evaluate to `null`.

In GraphQL, the `!` operator will act similarly. In the case that `name` does not exist, the query will return an error

> to the client rather than any data.

Reviewer 4: Can you expand here? My reading of this is that if the value of a field with `!` is null at runtime, then the GraphQL executor must treat that identically to the case that the field resolver raised an exception, including returning an entry in the top-level `errors` list and nulling out the parent field (which may need to then bubble up an error if that field was non-nullable in the schema, or also had `!` applied).

This would mean that in a query such as the following, `other_media_field` being null at runtime would cause the entire query to return effectively no data (just `{data: {media: null}, errors: [...]}`). This would prevent any part of the UI from rendering because a single field was missing for one part:

```
query MediaQuery {
  media(id: "...") {
     name
     thumbnail
     ...MediaFragment
  }
}
fragment MediaFragment on Media {
  other_media_field!
}
```

How are you thinking about handling cases such as this?

Reviewer 5: I read this as any field in the response that would be `null` and has a `!` on the field in the operation would cause the whole operation to return an error and no data. Even if that field is in some named fragment.

I think this would limit the usefulness extremely, basically to very simple queries. To take Yelp as an example, I don't think it'd be a good idea to error a search query if a single search result item failed to generate a star rating (assuming that's required to render a search result item).

It would be good to clarify what the behavior is if the `!` is in a named fragment.

Reviewer 3: At Netflix, we have implemented this on the server (using a directive syntax) and defined it to have identical semantics as the existing SDL [Non-Null behavior](https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullabilit). This has worked well.

I've proposed a [clarification](https://github.com/twof/graphql-spec/pull/5) to specify this.

Reviewer 4: @fotoetienne Thanks for adding the clarification, that confirms my interpretation and brings us back to my original question:

How are you thinking about the fact that a single fragment with a `!` field can cause data for other fragments (or in the extreme, an entire request) to be nulled out? As @captbaritone mentioned above, one of our goals / requirements when designing `@required` was that a "missing" required value for one part of a UI cannot cause data for other parts of the UI to be nulled out. The design here violates that constraint (it breaks fragment modularity).

Reviewer 3: @josephsavona That's a great question. Our current server logic will make a field act as Non-Null is there _any_ Non-Null designation for that field in the query. This query example:

```graphql
fragment Preview on Media {
  thumbnail(size: 32)

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

would essentially be interpreted by the server as:

```graphql
fragment Preview on Media {
  thumbnail(size: 32) @nonNull

  ... on Photo {
    thumbnail(size: 32) @nonNull
  }
}
```

It is at least non-ambiguous, but I can see how it is a challenge for fragment modularity.

Reviewer 4: @fotoetienne Thanks for clarifying. That seems like a severe limitation that would make applications significantly less robust than they can and arguably should be. How do you manage this in practice? Do you encourage fields with `!` to be given unique aliases so that you prevent removing data for other parts of a UI?

Reviewer 3: Short answer - it hasn't come up _for us_ yet, but it's an important edge-case to consider.

Long answer - Fragment modularity is particularly crucial for clients that rely heavily on components tied to individually defined Fragments that are compiled into large queries. We don't currently have many teams using that approach.

I agree that this is an important edge-case that we should think through some more. We want a design that will lead UI developers to fall into "the pit of success" rather than allowing surprising behavior like you've described.

Reviewer 3: An initial thought, probably flawed: reversing the server-defined logic might help. Only honor nonNull if _all_ references to that field specify it. This would, however, require client codegen to be aware of an entire query and only make types nonNull if all fragments in the query specify nonNull. Thoughts?

Reviewer 3: Making conflicting definitions between fragments be a validation error (like you did with @requires) is the safest approach. I like that. It's probably an acceptable limitation that you just can't have differing nullability for the same field.

Reviewer 5: @joesavona even unique aliases would still null out the object by bubbling up to the parent. It's really the same as any specific field being non-nullable in the server schema and ending up null which bubbles up.

This feels like a common enough need to call for a language feature. A single language feature would enable more unified public tooling around GraphQL.

### Make Schema Fields Non-Nullable Instead
Discussion on [this topic can be found here](https://medium.com/@calebmer/when-to-use-graphql-non-null-fields-4059337f6fc8)

### Write wrapper types that null-check fields
This is the alternative being used at some of the companies represented in this proposal for the time being.
It's quite labor intensive and the work is quite rote. It more or less undermines the purpose of
having code generation.

### Alternatives to `!`
#### `!!`
This would follow the precedent set by Kotlin.

### Make non-nullability apply recursively
For example, everything in this tree would be non-nullable
```graphql
query! {
  business(id: 4) {
    name
  }
}
```

## 🙅 Syntax Non-goals

This syntax consciously does not cover the following use cases:

> - **Default Values**

Reviewer 3: Defaults can be quite useful from a resilience standpoint. It is relevant to this proposal in that the `!` syntax would't be as amenable to specifying a default, whereas a directive (e.g. `@nonNull(default: false)`) would open us up to that.

Reviewer 6: Though replacing `!` with `! = “default”` like in variable defaults could work. This would either have to be evaluated client side (like in Relay) or the same default would have to apply to all references at the same location (whether from fragments or directly).

# RFC: Operation Level Nullability 

**Proposed by:**

- [Alex Reilly](<social or github link here>) - Yelp iOS
- [Liz Jakubowski](https://github.com/lizjakubowski) - Yelp iOS
- [Mark Larah](https://github.com/magicmark) - Yelp Web
- [Sanae Rosen](<social or github link here>) - Yelp Android
- [Stephen Spalding](https://github.com/fotoetienne) - Netflix GraphQL Server Infrastructure

> - [Wei Xue](<social or github link here>) - Yelp iOS

Reviewer 7: ```suggestion
- [Wei Xue](https://github.com/xuewei8910) - Yelp iOS
```

number, so that statement will evaluate to `null` rather than an integer if the string cannot be cast to an integer. If
you want to ensure that the statement does not return `null` you can instead write `Int("5")!`. If you do that, an
exception will be thrown if the statement would evaluate to `null`.

In GraphQL, the `!` operator will act similarly.

### `!`

We have chosen `!` because `!` is already being used in the GraphQL spec to indicate that a field is non-nullable.
Incidentally the same precedent exists in Swift (`!`) and Kotlin (`!!`) which both use similar syntax to indicate
"throw an exception if this value is `null`".

## Use cases

#### When a field is necessary to the function of the client

Expressing nullability in the query, as opposed to the schema, offers the client more flexibility and control over
whether or not an error is thrown.

There are cases where a field is `nullable`, but a feature that fetches the field will not function if it is `null`.
For example if you are trying to render an information page for a movie, you won't be able to do that if the name field

> of the movie is missing. In that case it would be preferable to fail as early as possible.

Reviewer 8: Here's one of the cases where this would be needed.

[GatsbyJS](https://www.gatsbyjs.com/) projects use GraphQL to handle build-time data dependencies, and provide a file system route API that can automatically build pages from specific types.

Creating a file called `src/pages/{BlogPost.slug}.jsx` creates pages from all BlogPost nodes and injects page context. `{ id: string, slug: string }`

user can use the page context as an argument to fetch additional data.

```graphql
# $id is automatically injected by Gatsby
query BlogPostQuery($id: String!) {
  # The type of blogPost is nullable, but the result is guaranteed to non-null by Gatsby
  blogPost(id: { eq: $id }) {
    title
    tags
  }
}
``` 

This can be done with additional static analysis in plugin layer. But the proposal can be a simple solution.

Even if It's not clear in production cases that can customize schema, but it seems a huge advantage for ecosystem players using SDL and code-generation.

Reviewer 2: Thank you for sharing another use case!

    name!
  }
}
```
would codegen to the following type on iOS.
```swift
struct GetBusinessNameQuery {
  let id: String

  struct Data {
    let business: Business?

    struct Business {
      /// Lack of `?` indicates that `name` will never be `null`
      let name: String
    }
  }
}
```

#### Non-nullable object with nullable fields
```graphql
query GetBusinessReviews {
  business(id: 4) {
    reviews! {
      rating
      text
    }
  }
}
```

## ✅ RFC Goals

- Non-nullable syntax that is based off of syntax that developers will already be familiar with
- Enable GraphQL codegen tools to generate more ergonomic types

## 🚫 RFC Non-goals

## 🗳️ Alternatives considered

### A `@nonNull` official directive

This solution offers the same benefits as the proposed solution. Since many GraphQL codegen tools already support the `@skip` and `@include` directives, this solution likely has a faster turnaround.

### A `@nonNull` custom directive

This is an alternative being used at some of the companies represented in this proposal for the time being.

While this solution simplifies some client-side logic, it does not meaningfully improve the developer experience for clients.

> * The cache implementations of GraphQL client libraries also need to understand the custom directive to behave correctly. Currently, when a client library caches a null field based on an operation without a directive, it will return the null field for another operation with this directive.

Reviewer 9: This will be true even with custom syntax. For composability of fragments, it would be worrisome if this new syntax made it impossible to have these two fragments in the same application:
```
fragment UserOkWithoutName on User {
  name
}
fragment UserMustHaveName on User {
  name!
}
```
With consistency, I could end up re-rendering a component built with `UserMustHaveName` based on a response fetched using only `UserOkWithoutName`.

And in fact it gets worse:
```
fragment BestFriendWithoutName on User {
  best_friend {
    id
  }
}
fragment BestFriendWithName on User {
  best_friend {
    id
    name!
  }
}
```
If my best friend, when I fetched `BestFriendWithName`, is `User:1`, and then suddenly, after fetching `BestFriendWithoutName` is `User:2`, I no longer am capable of rendering the component relying on `BestFriendWithName` (because `name` is *unset* on `User:2`). If the client library doesn't know what to do in that case, you can end up with consistency updates causing crashes.

Reviewer 2: As you said, the client library will need to understand this new syntax/directive. I am not familiar with GraphQL client implementations, so I am not sure how difficult my imagined behavior will be to implement. Doesn't this behavior solve the issue you outlined?

> If a GraphQL client finds that a component relies on `BestFriendWithName `, but the cached data doesn't have `name`, it should not return any of the cached data. To the client, it is like a state before fetching started. Depending on `fetch-policy` (to use Apollo's terminology), the client can start a new request to fetch the missing field data.

Reviewer 9: > Doesn't this behavior solve the issue you outlined?

That's how *Apollo* solves it, but that strategy isn't true universally. For instance that's not how Relay solves it. The client I work on *can't* solve it this way, as our consistency subscriptions are at the response level, for queries that include hundreds of fragments (whether that's a good choice or not is irrelevant: it's how things work for the largest GraphQL client at Facebook): if missing `name` meant the entire consistency update should not be served, whenever a product chose to mark their field with `!`, they could cause entire unrelated surfaces to lose consistency.

Another valid strategy would also be to return the data, but with `best_friend` being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for `@required(action: NONE)`). Or requiring the product to null check every piece of data, always, regardless of its nullability annotation.

Reviewer 2: Do you mean that Relay requires field selections on the same id and type to behave the same globally? If so, I can see how my suggested behavior won't work. If not:

> Another valid strategy would also be to return the data, but with best_friend being null: all the other data in the fragment/consistency subscription would still be delivered (this is Relay's strategy for @required(action: NONE))

This is what I imagined. The client would nullify up until the most immediate nullable parent.

Reviewer: @mjmahone thank you for the feedback, this is great for us to think about! ❤️  I think I see the problem you've laid out here, but let me know if I've misunderstood anything!

when a not-required field is fetched and returned as `null`, we could end up wiping away values in the cache that are required by other features.

for features using codegen types, the Swift/Kotlin/etc. properties corresponding to those "required" fields will be non-Optional. the value will be replaced by `null` in the cache -- which would result in a crash anytime those non-Optional properties are accessed.

is it feasible for the client to distinguish in the response between these two cases (e.g. perhaps via the `errors` array)?
* fields that are `null` due to an error, and
* fields that are `null` because they are also `null` in the database

in the former case, the client library could choose to not clear the existing value in the cache. from some experimentation, this doesn't seem to be how e.g. the Apollo iOS cache works with regular nullable fields -- if a `null` value is returned, it seems to replace any value that was previously in the cache (I wasn't able to verify that this was also the case when an error occurs on that field).

is there any other way for the client library to know what to do in this case?
